cpp/src/arrow/util/async_generator.h

```diff
@@ -177,6 +330,124 @@ class TransformingGenerator {
   std::shared_ptr<TransformingGeneratorState> state_;
 };
 
+/// \brief Transforms an async generator using a transformer function returning a new
```

Comment:
It seems this + the queue files could be removed in favor of the version in #9533?

Reply:
Hmm, I'm not sure what you mean here.  The version in #9533 still relies on the transforming generator.

Reply:
Sorry, I mean that since nothing in this PR uses this or the queue, it could be removed from this PR and reviewed solely in the context of the other PR.

Reply:
Ah, right.  All of these build on each other.  So hopefully these changes will magically just go away once #9533 is merged.  I think it would be too much work for me to keep track of the branching otherwise.  The next PR 11887 needs both these changes and the changes from #9533 so I wouldn't be saving too much by splitting at the moment.